src/eth.rs

```diff
@@ -173,14 +308,180 @@ impl VitalikProvider {
+    ///     Some(15), // Timeout for all Ethereum interactions
+    /// ).build().unwrap();
+    /// ```
+    /// # Errors
+    pub fn new(
+        api_url: String,
+        api_key: String,
+        chain_id: Option<u64>,
+        private_key: Option<String>,
+        contract_address: Option<Address>,
+        timeout: Option<u64>,
+    ) -> Self {
+        Self {
+            api_url,
+            api_key,
+            chain_id,
+            private_key,
+            contract_address,
+            timeout
+        }
+    }
+
+    pub fn build(&self) -> Result<EthClient, Error> {
+        // Determine an API URL and Initialize the Provider
+        let url = format!("{}{}", self.api_url, self.api_key);
+        let provider = Provider::<Http>::try_from(url).expect("Failed to create provider");
+        // Get the Chain ID. If None, set to 1
+        let chain_id = self.chain_id.unwrap_or(1);
+        // Check if we have a private key to set up a Signer
+        let signer = if let Some(private_key) = &self.private_key {
+            let mut wallet = private_key.parse::<LocalWallet>().expect("Failed to parse private key");
+            wallet = wallet.with_chain_id(chain_id);
+            Some(SignerMiddleware::new(provider.clone(), wallet))
+        }
+        else {
+            None
+        };
+        // Check if we have a contract address to set up a Contract
+        let contract = if let Some(contract_address) = &self.contract_address {
+            let abi: Abi = serde_json::from_str(&BANYAN_ABI_STR_REF).expect("Failed to parse ABI");
+            let contract = Contract::new(
+                contract_address.clone(), abi, provider.clone()
+            );
+            Some(contract)
+        }
+        else {
+            None
+        };
+        // Determine the timeout as a Duration in seconds, assign default if not provided
+        let timeout = Duration::from_secs(self.timeout.unwrap_or(15));
+        Ok(EthClient {
+            provider,
+            chain_id,
+            signer,
+            contract,
+            timeout,
+        })
+    }
+}
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test]
+    /// Test Init a new eth client from the environment.
+    /// The environment variables for all fields must be set for this test to pass
+    async fn eth_client() {
+        // Init a new EthClient with our environment variables
+        let eth_client = EthClientBuilder::default().build().unwrap();
+        if !eth_client.has_signer() {
+            panic!("No signer available!");
+        }
+        if !eth_client.has_contract() {
+            panic!("No contract available!");
+        }
+        // Try and get the current block number
+        let block_num = eth_client.get_latest_block_num().await.unwrap();
+        println!("Latest Block Number: {}", block_num.0);
+        return;
+    }
+
+    #[tokio::test]
+    /// Test sending a deal Proposal
```

Comment:
good, but add more tests?

Reply:
willl do -- once I got working with the real contracts working.

Reply:
This does not work -- testing with an empty contract I deployed. Working on it.

Reply:
Got deal Proposal posting working (assuming it is configured properly). Working on detokenizing onchaindeals with the getter x defined and then I can make this test more robust by pulling the created deal.